### PR TITLE
Update runtime dependency packages to latest from maintenance-packages

### PR DESCRIPTION
**Pull request by carlossanlop**
Unblocks https://github.com/dotnet/runtime/pull/108806

We have generated new preview versions of some OOB packages that have been recently migrated to the maintenance-packages repo. We need to include them in sbrp so that we can update runtime's `main` dependencies to those preview versions without breaking source build.

The packages that we want to update in runtime are:
- Microsoft.Bcl.HashCode
- System.Buffers
- System.Memory
- System.Numerics.Vectors
- System.Runtime.CompilerServices.Unsafe
- System.Threading.Tasks.Extensions

I used the generate.sh script to generate these files, per the instructions.

For additional details, see the decided plan here: https://github.com/dotnet/maintenance-packages/pull/140#issuecomment-2444383987

cc @ericstj @ViktorHofer @MichaelSimons

### CODE REVIEW

**Comment by MichaelSimons** [comment]
Per the [instructions for adding new packages](https://github.com/dotnet/source-build-reference-packages/blob/main/README.md#workflow): 

Add DependencyPackageProjects for all new projects in the [eng/Build.props](https://github.com/dotnet/source-build-reference-packages/blob/main/eng/Build.props#L9) in the correct dependency order.

**Comment by carlossanlop** [comment]
Done. Sorry, I missed that line in the instructions.

File: eng/Build.props [hunk 1]
```diff
@@ -36,6 +37,11 @@
     <DependencyPackageProjects Include="$(RepoRoot)src\referencePackages\src\**\Nuget.Commands.6.11.0.csproj" />
     <DependencyPackageProjects Include="$(RepoRoot)src\referencePackages\src\**\System.Text.Json.8.0.5.csproj" />
     <DependencyPackageProjects Include="$(RepoRoot)src\textOnlyPackages\src\**\Microsoft.NetCore.Platforms.3.1.0.csproj" />
+    <DependencyPackageProjects Include="$(RepoRoot)src\referencePackages\src\**\System.Buffers.4.6.0-preview.1.24529.4.csproj" />
+    <DependencyPackageProjects Include="$(RepoRoot)src\referencePackages\src\**\System.Memory.4.6.0-preview.1.24529.4.csproj" />
```
Comment (ViktorHofer) [inline eng/Build.props:41]: System.Memory depends on Buffers, Vectors and Unsafe and therefore must be listed after those.

**Comment by carlossanlop** [comment]
I noticed Bcl.AsyncInterfaces is not added to Build.props, is that expected? (I did not add that one, it was already there).

**Comment by carlossanlop** [comment]
Oh - We only add to DependencyPackageProjects those projects that are dependency of others. We don't add all that we add. There are many other packages in the referencePackages folder that are not in Build.props.

File: eng/Build.props [hunk 2]
```diff
@@ -36,6 +36,11 @@
     <DependencyPackageProjects Include="$(RepoRoot)src\referencePackages\src\**\Nuget.Commands.6.11.0.csproj" />
     <DependencyPackageProjects Include="$(RepoRoot)src\referencePackages\src\**\System.Text.Json.8.0.5.csproj" />
     <DependencyPackageProjects Include="$(RepoRoot)src\textOnlyPackages\src\**\Microsoft.NetCore.Platforms.3.1.0.csproj" />
+    <DependencyPackageProjects Include="$(RepoRoot)src\referencePackages\src\**\System.Buffers.4.6.0-preview.1.24529.4.csproj" />
+    <DependencyPackageProjects Include="$(RepoRoot)src\referencePackages\src\**\System.Numerics.Vectors.4.6.0-preview.1.24529.4.csproj" />
+    <DependencyPackageProjects Include="$(RepoRoot)src\referencePackages\src\**\System.Runtime.CompilerServices.Unsafe.6.1.0-preview.1.24529.4.csproj" />
+    <DependencyPackageProjects Include="$(RepoRoot)src\referencePackages\src\**\System.Memory.4.6.0-preview.1.24529.4.csproj" />
+    <DependencyPackageProjects Include="$(RepoRoot)src\referencePackages\src\**\System.Threading.Tasks.Extensions.4.6.0-preview.1.24529.4.csproj" />
```
Comment (carlossanlop) [inline eng/Build.props:43]: Memory depends on Buffers, Vectors and Unsafe.
Threading.Tasks.Extensions depends on Unsafe.

**Comment by carlossanlop** [comment]
Oh ok got it. Then I'll re-add the one I just removed, Bcl.HashCode.

**Review by ellahathaway — APPROVED** [review]
LGTM.

The only thing that caught my eye was the preview versions, but I see that using previous versions here was approved by @MichaelSimons (see https://github.com/dotnet/maintenance-packages/pull/140#issuecomment-2444383987) and also mentioned in the PR description.

**Comment by carlossanlop** [comment]
This is all going into main, we want to test these preview versions in runtime. I am currently in the process of stabilizing the m-p repo to generate the official GA versions of these packages. If needed, I'll be happy to generate the reference packages for GA too.

File: NuGet.config [hunk 3]
```diff
@@ -5,6 +5,7 @@
   </fallbackPackageFolders>
   <packageSources>
     <clear />
+    <add key="dotnet-libraries" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-libraries/nuget/v3/index.json" />
```
Comment (ViktorHofer) [inline NuGet.config:8]: Should this feed get checked in?

**Review by ViktorHofer — APPROVED** [review]

File: NuGet.config (hunk 3 above)
Comment (MichaelSimons) [reply NuGet.config:8]: Yes it should - as long at the checked in packages come from this feed it should be checked in to support regen.  This should get cleaned up once the preview maintenance packages are removed and the feed is no longer required.

Comment (carlossanlop) [reply NuGet.config:8]: The generate.sh script fails to find the packages without this feed.

**Pull request merged**